cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/provisioner/dataproc/DefaultDataprocClientFactory.java

```diff
@@ -0,0 +1,67 @@
+/*
+ *  Copyright © 2022 Cask Data, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ *  use this file except in compliance with the License. You may obtain a copy of
+ *  the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ *  License for the specific language governing permissions and limitations under
+ *  the License.
+ */
+
+package io.cdap.cdap.runtime.spi.provisioner.dataproc;
+
+import com.google.api.gax.core.CredentialsProvider;
+import com.google.api.gax.core.FixedCredentialsProvider;
+import com.google.api.services.compute.Compute;
+import com.google.cloud.dataproc.v1.ClusterControllerClient;
+import com.google.cloud.dataproc.v1.ClusterControllerSettings;
+import io.cdap.cdap.runtime.spi.ssh.SSHPublicKey;
+
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import java.util.Optional;
+import javax.annotation.Nullable;
+
+/**
+ * Creates DataprocClients.
+ */
+public class DefaultDataprocClientFactory implements DataprocClientFactory {
+  private final ComputeFactory computeFactory;
+
+  public DefaultDataprocClientFactory() {
+    this(new GoogleComputeFactory());
+  }
+
+  public DefaultDataprocClientFactory(ComputeFactory computeFactory) {
+    this.computeFactory = computeFactory;
+  }
+
+  @Override
+  public DataprocClient create(DataprocConf conf, boolean requireSSH,
+                               @Nullable SSHPublicKey sshPublicKey) throws IOException, GeneralSecurityException {
+    ClusterControllerClient clusterControllerClient = getClusterControllerClient(conf);
+    Compute compute = computeFactory.createCompute(conf.getComputeCredential(), conf.getComputeConnectionTimeout(),
```

Comment:
This is relatively cheap, correct? As we create it even if it's not needed.

Reply:
It should be, I believe the most expensive part is loading the credential. 

I'll spend a little time seeing if it's a small change to only create this when needed.